doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+
+The MVP for PostgreSQL sources will not support unmaterialized sources. It's not
+fundamentally impossible to have multiple instances of a PostgreSQL source, but
+each instance of a PostgreSQL source is prohibitively expensive. Forcing
+PostgreSQL sources to be materialized is an easy way to ensure that only one
+instance of a PostgreSQL source is ever created.
+
+Roughly speaking, PostgreSQL's replication protocol works per database, while
+a PostgreSQL source in Materialize corresponds to just one table in that
+database. If you were to instantiate a PostgreSQL source twice in Materialize,
+you'd wind up streaming two separate copies of the entire database to
+Materialize, which is certain to be surprising to users.
+
+With dynamic resource sharing, creating a new instance of a PostgreSQL source
+would instead backfill the missing data for just the one table, then cut over to
+the already existing replication stream. This will be cheap enough that we'll
+be comfortable allowing PostgreSQL sources to be unmaterialized.
+
+### Discussion
+
+The major downside of this approach is the implementation complexity. Cutting
+source instances over from a backfill resource to the steady-state resource will
+require a lot of fiddly logic and testing.
+
+My contention is that this complexity is well worth it for the user experience
+it provides. Expecting users to predict exactly which fields they'll need from a
+source is unrealistic. Asking them to recreate an entire stack of views, as
+described in the [workaround](#todays-workaround) is acceptable in the short
+term but unfriendly in the long term. But letting users interactively create the
+sources and views they want, while converging on the optimal implementation
+nonetheless? That's the magic experience we strive for.
+
+Also, the lessons learned here will be directly applicable to our future plans
+for multi-query optimization. Dynamically sharing resources in sources is a
+simplified version of dynamically sharing resources (i.e., arrangements) between
+views.
+
+## Alternatives
+
+### Bulk creation
+
+We'd previously talked about using SQL transactions to permit bulk creation
+of set of sources and views, like so:
+
+```sql
+BEGIN;
+CREATE SOURCE kafka_src FROM KAFKA BROKER '...' TOPIC 'top' FORMAT AVRO ...
+CREATE MATERIALIZED VIEW view1 AS SELECT col1 FROM kafka_src
+CREATE MATERIALIZED VIEW view2 AS SELECT col2 FROM kafka_src
+COMMIT;
+```
+
+When the transaction commits, Materialize could look holistically at the
+objects within, and realize that only one new source instance is necessary.
+
+But this turns out not to be any more flexible than the [current
+workaround](#todays-workaround). You need to be willing to declare all of your
+sources and views in one shot, and that's not always feasible.
+
+We might, however, want to support this feature for savvy users though as a
+minor optimization for dynamic resource sharing. Dynamic resource sharing only
```

Comment:
For postgres sources specifically it's actually beneficial to create the sources one by one as the backfill consists of a COPY statement. The replication implementation in pg uses one connection per table for the initial sync too before it cuts over to the main replication stream.